### PR TITLE
Add support for Node 18

### PR DESCRIPTION
**Pull request by ejizba**
Related to https://github.com/Azure/azure-functions-nodejs-worker/issues/562

azurefunctions.azurecr.io/azure-functions/4/node:ejtest-node18: 1.0 GB
<details>
<summary>dive results</summary>
image1
</details>

azurefunctions.azurecr.io/azure-functions/4/node:ejtest-node18-slim: 1.0 GB
<details>
<summary>dive results</summary>
image2
</details>

azurefunctions.azurecr.io/azure-functions/4/node:ejtest-node18-appservice: 857 mb

<details>
<summary>dive results</summary>
image3
</details>

### CODE REVIEW

File: host/4/bullseye/amd64/node/node18/node18-composite.template [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+RUN apt-get update && \
+    apt-get install -y --no-install-recommends openssh-server dialog && \
+    echo "root:Docker!" | chpasswd
+
+# This is current not working with the .NET 6.0 image based on bullseye. Tracking here: https://github.com/Azure/azure-functions-docker/issues/451
+# Chrome headless dependencies
+# https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md#chrome-headless-doesnt-launch-on-unix
+# RUN apt-get install -y xvfb gconf-service libasound2 libatk1.0-0 libc6 libcairo2 libcups2 \
+#    libdbus-1-3 libexpat1 libfontconfig1 libgbm1 libgcc1 libgconf-2-4 libgdk-pixbuf2.0-0 libglib2.0-0 \
```
Comment (pragnagopa) [inline host/4/bullseye/amd64/node/node18/node18-composite.template]: please go ahead and delete commented code

**Comment by pragnagopa** [comment]
Please open follow up issues to update ReadMe with once the node18 images are available. Also, open follow up issues to include node18 images in mcr docs - links to repos are listed in https://github.com/Azure/azure-functions-docker#mcr-docs

**Comment by pragnagopa** [comment]
See PR description for https://github.com/Azure/azure-functions-docker/pull/632 as example - add image sizes specifically for the app service images - tagging @CooperLink to guide the process he follows to verify sizes and docker  layer sharing

We do have a tracking item to verify image sizes as part of the PR builds - for now, please collect this info manually

**Comment by ejizba** [comment]
First of all, Node 18 support in the worker [will be included](https://github.com/Azure/azure-functions-host/pull/8467) in v4.7.0 of the host. Cutoff date for changes was today, so the github tag should be created soon and then I should be able to get the build to pass.

> Please open follow up issues to update ReadMe with once the node18 images are available.

I already updated the README in this PR. Do I need to revert that change for now and add to the README later?

> Also, open follow up issues to include node18 images in mcr docs - links to repos are listed in https://github.com/Azure/azure-functions-docker#mcr-docs

Done: https://github.com/Azure/azure-functions-docker/issues/717

> add image sizes specifically for the app service images

Done - added to PR description above

File: host/4/bullseye/amd64/node/node18/node18-composite.template [hunk 2]
```diff
@@ -0,0 +1,54 @@
+FROM mcr.microsoft.com/dotnet/runtime-deps:6.0
+ARG HOST_VERSION
+
+COPY --from=runtime-image [ "/FuncExtensionBundles", "/FuncExtensionBundles" ]
+COPY sshd_config /etc/ssh/
+COPY start.sh /azure-functions-host/
+COPY --from=runtime-image [ "/azure-functions-host", "/azure-functions-host" ]
+COPY --from=runtime-image [ "/workers/node", "/azure-functions-host/workers/node" ]
```
Comment (CooperLink) [inline host/4/bullseye/amd64/node/node18/node18-composite.template:8]: Verifying that Layer sharing is good here. There is only one set of node worker files right? The code just depends on which Node Lang Version is available? If there is more than one node worker defined in Host let me know and I can work on optimizing these images.

File: host/4/bullseye/amd64/node/node18/node18-composite.template [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    apt-get install -y nodejs
+
+# This is current not working with the .NET 6.0 image based on bullseye. Tracking here: https://github.com/Azure/azure-functions-docker/issues/451
+# Chrome headless dependencies
+# https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md#chrome-headless-doesnt-launch-on-unix
+#RUN apt-get install -y xvfb gconf-service libasound2 libatk1.0-0 libc6 libcairo2 libcups2 \
+#    libdbus-1-3 libexpat1 libfontconfig1 libgbm1 libgcc1 libgconf-2-4 libgdk-pixbuf2.0-0 libglib2.0-0 \
+#    libgtk-3-0 libnspr4 libpango-1.0-0 libpangocairo-1.0-0 libstdc++6 libx11-6 libx11-xcb1 libxcb1 \
+#    libxcomposite1 libxcursor1 libxdamage1 libxext6 libxfixes3 libxi6 libxrandr2 libxrender1 libxss1 \
+#    libxtst6 ca-certificates fonts-liberation libappindicator1 libnss3 lsb-release xdg-utils wget
```
Comment (CooperLink) [inline host/4/bullseye/amd64/node/node18/node18-composite.template]: Remove comment block

File: host/4/bullseye/amd64/node/node18/node18-core-tools.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+# The javascript-node image includes a non-root user with sudo access. Use the "remoteUser"
+# property in devcontainer.json to use it. On Linux, the container user's GID/UIDs
+# will be updated to match your local UID/GID (when using the dockerFile property).
+# See https://aka.ms/vscode-remote/containers/non-root-user for details.
+ARG USERNAME=node
+ARG USER_UID=1000
+ARG USER_GID=$USER_UID
+
+# Configure apt and install packages
+RUN apt-get update \
```
Comment (CooperLink) [inline host/4/bullseye/amd64/node/node18/node18-core-tools.Dockerfile:20]: @pragnagopa  Do we want new languages to onboard a core-tools image in this repo?

**Review by CooperLink — DISMISSED** [review]
This LGTM only one nit on removing a repeated comment block. Thank you for your contribution.

File: host/4/bullseye/amd64/node/node18/node18-composite.template (hunk 2 above)
Comment (ejizba) [reply host/4/bullseye/amd64/node/node18/node18-composite.template:8]: Do you have an example of a language that has more than one worker in the host? Are you talking about PowerShell how they have a 7.0 and a 7.2 worker? If so, we are _not_ like PowerShell.

We use the same worker in v4 for all supported node versions (v14, v16 & now v18). We have a different worker in v3 of the host, but I don't think that's what you're asking about - and it's still just one worker used for all of v3.

**Comment by ejizba** [comment]
@CooperLink @pragnagopa any idea when we can get this merged/released? I assume this missed 4.7.0, but can we make it for 4.8.0?

Same question for the private repo PR

File: host/4/bullseye/amd64/node/node18/node18-composite.template (hunk 2 above)
Comment (CooperLink) [reply host/4/bullseye/amd64/node/node18/node18-composite.template:8]: Another example is Python. They produce different flavors of worker per language version and I copy over just one of the workers like this https://github.com/Azure/azure-functions-docker/blob/dev/host/3.0/buster/amd64/python/python36/python36-composite.template#L69

**Review by CooperLink — APPROVED** [review]
This looks good to me. We need to verify that Platform currently can support the new node images. It should already.

File: host/4/bullseye/amd64/node/node18/node18-core-tools.Dockerfile (hunk 4 above)
Comment (pragnagopa) [reply host/4/bullseye/amd64/node/node18/node18-core-tools.Dockerfile:20]: Unfortunately, yes until https://github.com/Azure/azure-functions-core-tools/issues/2748 is resolved.

**Review by pragnagopa — APPROVED** [review]

**Comment by pragnagopa** [comment]
> @CooperLink @pragnagopa any idea when we can get this merged/released? I assume this missed 4.7.0, but can we make it for 4.8.0?

That's right. Merging the PR now. Will be part of next host release - cc @VpOfEngineering for exact timelines

**Pull request merged**